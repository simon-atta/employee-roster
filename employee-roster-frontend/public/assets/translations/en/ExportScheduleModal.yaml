export: Export
exportSchedule: Export Schedule
fromDate: From Date
toDate: To Date
forSpots: For Spots
selectSpots: Select Spots