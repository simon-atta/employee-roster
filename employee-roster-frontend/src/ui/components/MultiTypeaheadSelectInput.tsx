/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import Select from 'react-select';

export interface MultiTypeaheadSelectProps<T> {
  emptyText: string;
  options: T[];
  value: T[];
  optionToStringMap: (option: T) => string;
  onChange: (selected: T[]) => void;
  optional?: boolean;
  noClearButton?: boolean;
  autoSize?: boolean;
}

export default class MultiTypeaheadSelectInput<T> extends React.Component<MultiTypeaheadSelectProps<T>> {
  constructor(props: MultiTypeaheadSelectProps<T>) {
    super(props);
    this.onSelect = this.onSelect.bind(this);
  }

  onSelect(selections: { value: T }[]|null) {
    if (selections !== null) {
      this.props.onChange(selections.map(s => s.value));
    } else {
      this.props.onChange([]);
    }
  }

  render() {
    const { optionToStringMap, emptyText, optional, options, autoSize } = this.props;
    const selectOptions = this.props.options.map(o => ({ value: o }));
    // Why map to string first?
    // Clone of objects are not the same, and sometimes we are passed
    // clones instead of the real thing, and "===" and "==" are
    // both identity/is comparators on objects, so we compare their
    // "toString" values to check for equality
    const selected = selectOptions.filter(o => this.props.value
      .find(item => optionToStringMap(item) === optionToStringMap(o.value)) !== undefined);
    const parentStyle: React.CSSProperties = {
      position: 'relative',
    };
    if (autoSize === undefined || autoSize) {
      const maxLabelLength = Math.max(emptyText.length, ...options.map(optionToStringMap).map(s => s.length));
      parentStyle.width = `${maxLabelLength + 2}em`;
    }

    return (
      <div style={parentStyle}>
        <Select
          aria-label={emptyText}
          styles={{
            menuPortal: provided => ({
              ...provided,
              zIndex: 9999999,
            }),
          }}
          onChange={this.onSelect}
          defaultValue={selected}
          value={selected}
          selected={selected}
          placeholder={emptyText}
          getOptionLabel={o => optionToStringMap(o.value)}
          getOptionValue={o => optionToStringMap(o.value)}
          options={selectOptions}
          menuPortalTarget={document.body}
          isMulti
          isClearable={optional}
        />
      </div>
    );
  }
}
