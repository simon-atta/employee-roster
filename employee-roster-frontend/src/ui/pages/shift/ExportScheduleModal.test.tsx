/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import DatePicker from 'react-datepicker';
import { Spot } from 'domain/Spot';
import moment from 'moment';
import { useTranslation } from 'react-i18next';
import { Modal } from '@patternfly/react-core';
import MultiTypeaheadSelectInput from 'ui/components/MultiTypeaheadSelectInput';
import { ExportScheduleModal } from './ExportScheduleModal';

describe('Export Schedule Modal', () => {
  it('should render correctly when closed', () => {
    const exportScheduleModal = shallow(<ExportScheduleModal
      {...baseProps}
      isOpen={false}
    />);
    expect(toJson(exportScheduleModal)).toMatchSnapshot();
  });

  it('should render correctly when opened', () => {
    const exportScheduleModal = shallow(<ExportScheduleModal
      {...baseProps}
    />);
    expect(toJson(exportScheduleModal)).toMatchSnapshot();
  });

  it('should export the schedule with default from and to date with the full spot list by default', () => {
    const exportScheduleModal = shallow(<ExportScheduleModal
      {...baseProps}
    />);
    const urlButton = shallow(exportScheduleModal.find(Modal).prop('actions')[1]);
    expect(urlButton.prop('href'))
      .toEqual(getExportUrlFor(baseProps.defaultFromDate, baseProps.defaultToDate, baseProps.spotList));
  });

  it('should export to _blank if spot list is empty', () => {
    const exportScheduleModal = shallow(<ExportScheduleModal
      {...baseProps}
      spotList={[]}
    />);
    const urlButton = shallow(exportScheduleModal.find(Modal).prop('actions')[1]);
    expect(urlButton.prop('href')).toEqual('_blank');
  });

  it('should export with new params if user changes fields', () => {
    const exportScheduleModal = shallow(<ExportScheduleModal
      {...baseProps}
    />);

    const newFromDate = moment('2020-08-01').toDate();
    exportScheduleModal.find(DatePicker).filter('[aria-label="Trans(i18nKey=fromDate)"]')
      .simulate('change', newFromDate);

    const newToDate = moment('2021-07-02').toDate();
    exportScheduleModal.find(DatePicker).filter('[aria-label="Trans(i18nKey=toDate)"]')
      .simulate('change', moment(newToDate).toDate());

    const otherSpot: Spot = {
      tenantId: 1,
      id: 2,
      version: 0,
      name: 'Spot',
      requiredSkillSet: [],
    };

    exportScheduleModal.find(MultiTypeaheadSelectInput).simulate('change', [spot, otherSpot]);
    const urlButton = shallow(exportScheduleModal.find(Modal).prop('actions')[1]);
    expect(urlButton.prop('href'))
      .toEqual(getExportUrlFor(newFromDate, newToDate, [spot, otherSpot]));
  });
});

function getExportUrlFor(fromDate: Date, toDate: Date, spotList: Spot[]): string {
  return `${process.env.REACT_APP_BACKEND_URL}/rest/tenant/1/roster/shiftRosterView/excel?`
                    + `startDate=${moment(fromDate).format('YYYY-MM-DD')}&`
                    + `endDate=${moment(toDate).format('YYYY-MM-DD')}&spotList=${spotList.map(s => s.id).join(',')}`;
}

const spot: Spot = {
  tenantId: 1,
  id: 2,
  version: 0,
  name: 'Spot',
  requiredSkillSet: [
    {
      tenantId: 0,
      id: 3,
      version: 0,
      name: 'Skill',
    },
  ],
};

const baseProps = {
  ...useTranslation('ExportScheduleModal'),
  tReady: true,
  isOpen: true,
  onClose: jest.fn(),
  defaultFromDate: moment('2018-07-01').toDate(),
  defaultToDate: moment('2018-07-07').toDate(),
  tenantId: 1,
  spotList: [spot],
};
