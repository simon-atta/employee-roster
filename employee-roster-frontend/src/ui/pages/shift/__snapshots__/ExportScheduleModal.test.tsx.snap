// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Schedule Modal should render correctly when closed 1`] = `
<Modal
  actions={
    Array [
      <Button
        aria-label="Close Modal"
        onClick={[MockFunction]}
        variant="tertiary"
      >
        Trans(i18nKey=close)
      </Button>,
      <a
        className="pf-c-button pf-m-primary"
        download={true}
        href="/rest/tenant/1/roster/shiftRosterView/excel?startDate=2018-07-01&endDate=2018-07-07&spotList=2"
        onClick={[Function]}
      >
        Trans(i18nKey=export)
      </a>,
    ]
  }
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={false}
  onClose={[MockFunction]}
  ouiaSafe={true}
  showClose={true}
  title="Trans(i18nKey=exportSchedule)"
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <Form
    id="modal-element"
    onSubmit={[Function]}
  >
    <InputGroup>
      <Label>
        Trans(i18nKey=fromDate)
      </Label>
      <t
        allowSameDay={false}
        aria-label="Trans(i18nKey=fromDate)"
        dateFormat="MM/dd/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2018-07-01T00:00:00.000Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      />
    </InputGroup>
    <InputGroup>
      <Label>
        Trans(i18nKey=toDate)
      </Label>
      <t
        allowSameDay={false}
        aria-label="Trans(i18nKey=toDate)"
        dateFormat="MM/dd/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2018-07-07T00:00:00.000Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      />
    </InputGroup>
    <InputGroup>
      <Label>
        Trans(i18nKey=forSpots)
      </Label>
      <MultiTypeaheadSelectInput
        aria-label="Trans(i18nKey=forSpots)"
        emptyText="Trans(i18nKey=selectSpots)"
        onChange={[Function]}
        optionToStringMap={[Function]}
        options={
          Array [
            Object {
              "id": 2,
              "name": "Spot",
              "requiredSkillSet": Array [
                Object {
                  "id": 3,
                  "name": "Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 1,
              "version": 0,
            },
          ]
        }
        value={
          Array [
            Object {
              "id": 2,
              "name": "Spot",
              "requiredSkillSet": Array [
                Object {
                  "id": 3,
                  "name": "Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 1,
              "version": 0,
            },
          ]
        }
      />
    </InputGroup>
  </Form>
</Modal>
`;

exports[`Export Schedule Modal should render correctly when opened 1`] = `
<Modal
  actions={
    Array [
      <Button
        aria-label="Close Modal"
        onClick={[MockFunction]}
        variant="tertiary"
      >
        Trans(i18nKey=close)
      </Button>,
      <a
        className="pf-c-button pf-m-primary"
        download={true}
        href="/rest/tenant/1/roster/shiftRosterView/excel?startDate=2018-07-01&endDate=2018-07-07&spotList=2"
        onClick={[Function]}
      >
        Trans(i18nKey=export)
      </a>,
    ]
  }
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[MockFunction]}
  ouiaSafe={true}
  showClose={true}
  title="Trans(i18nKey=exportSchedule)"
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <Form
    id="modal-element"
    onSubmit={[Function]}
  >
    <InputGroup>
      <Label>
        Trans(i18nKey=fromDate)
      </Label>
      <t
        allowSameDay={false}
        aria-label="Trans(i18nKey=fromDate)"
        dateFormat="MM/dd/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2018-07-01T00:00:00.000Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      />
    </InputGroup>
    <InputGroup>
      <Label>
        Trans(i18nKey=toDate)
      </Label>
      <t
        allowSameDay={false}
        aria-label="Trans(i18nKey=toDate)"
        dateFormat="MM/dd/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2018-07-07T00:00:00.000Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      />
    </InputGroup>
    <InputGroup>
      <Label>
        Trans(i18nKey=forSpots)
      </Label>
      <MultiTypeaheadSelectInput
        aria-label="Trans(i18nKey=forSpots)"
        emptyText="Trans(i18nKey=selectSpots)"
        onChange={[Function]}
        optionToStringMap={[Function]}
        options={
          Array [
            Object {
              "id": 2,
              "name": "Spot",
              "requiredSkillSet": Array [
                Object {
                  "id": 3,
                  "name": "Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 1,
              "version": 0,
            },
          ]
        }
        value={
          Array [
            Object {
              "id": 2,
              "name": "Spot",
              "requiredSkillSet": Array [
                Object {
                  "id": 3,
                  "name": "Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 1,
              "version": 0,
            },
          ]
        }
      />
    </InputGroup>
  </Form>
</Modal>
`;
